New doc for TeamsSSOProvider, new Get started for SSO Teams Tab, Updated Teams Tab (#13781)
microsoftgraph/microsoft-graph-docs: concepts/toolkit/get-started/build-a-microsoft-teams-sso-tab.md

---
title: "Build a Microsoft Teams SSO tab with the Microsoft Graph Toolkit"
description: "Get started building a Microsoft Teams SSO tab using the Microsoft Graph Toolkit."
localization_priority: Normal
author: simonagren
---

> # Build a Microsoft Teams SSO tab with the Microsoft Graph Toolkit

Reviewer: ```suggestion
# Build a Microsoft Teams Single Sign-On (SSO) tab with the Microsoft Graph Toolkit
```

8. Add components.
9. Test your app.

## Add the Microsoft Graph Toolkit

You can use the Microsoft Graph Toolkit in your application by referencing the loader directly (via unpkg) or by installing the npm package. To use the Toolkit, you will also need the [Microsoft Teams SDK](/javascript/api/overview/msteams-client?view=msteams-client-js-latest&preserve-view=true#using-the-sdk).

### Use via mgt-loader
To use the Toolkit and the Teams SDK via the loaders, add the reference in a script to your code:

```html
<!-- Microsoft Teams sdk must be referenced before the toolkit -->
<script src="https://unpkg.com/@microsoft/teams-js/dist/MicrosoftTeams.min.js" crossorigin="anonymous"></script>
<script src="https://unpkg.com/@microsoft/mgt/dist/bundle/mgt-loader.js"></script>
```

### Use via npm (ES6 modules)
Using the Toolkit via ES6 modules will give you full control of the bundling process and allow you to bundle only the code you need for your application. To use the ES6 modules, add the npm packages for both the Toolkit and the Microsoft Teams SDK to your project:

```cmd
npm install @microsoft/teams-js @microsoft/mgt-element @microsoft/mgt-teams-msal2-provider

> ```

Reviewer: Can we format this into tabs? (example here: https://github.com/microsoftgraph/microsoft-graph-docs/blob/main/concepts/toolkit/get-started/build-a-web-app.md). I can't add a suggestion with the code because it formats weirdly :(

## Create the auth popup page

> In order to allow users to sign in, you need to provide a URL that the Teams app will open in a popup to follow the authentication flow. The URL needs to be in your domain, and all this page needs to do is call the `TeamsSSOProvider.handleAuth()` method.

Reviewer 2: ```suggestion
If you need a user to consent to a permission, you need to provide a URL that the Teams app will open in a popup to follow the authentication flow. The URL needs to be in your domain, and all this page needs to do is call the `TeamsSSOProvider.handleAuth()` method.
```

You can do this by adding a new `auth.html` file (which should be at the same level as `index.html`) and adding the following code:

```html
<!DOCTYPE html>
<html>
  <head>
    <script src="https://unpkg.com/@microsoft/teams-js/dist/MicrosoftTeams.min.js" crossorigin="anonymous"></script>
    <script src="https://unpkg.com/@microsoft/mgt/dist/bundle/mgt-loader.js"></script>
  </head>

  <body>
    <script>
      mgt.TeamsSSOProvider.handleAuth();
    </script>
  </body>
</html>
```

> ## Creating an app/client ID

Reviewer 2: I recommend including the steps to create app id here directly to avoid sending the user somewhere else

Have a look at the section [Creating the AAD registration](https://github.com/microsoftgraph/microsoft-graph-toolkit/blob/main/samples/teams-sso-node/readme.md#creating-the-aad-registration) from the `readme` of the [Teams SSO Node Sample](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/samples/teams-sso-node). This involves registering the AAD application and configuring it for SSO.

## Create the backend

The backend could be any backend that enables exchanging the Microsoft Teams authentication token with a token that can be used to call Microsoft Graph via the [on-behalf-of flow](https://docs.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow). 

Have a look at the [Teams SSO Node Sample](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/samples/teams-sso-node) for reference.

### Server folder
At the app root, create a folder named **server** (the same level as the **tabs** folder). Inside of the folder, create and copy the values from the sample:

1. `package.json`
2. `tsconfig.ts`
3. `.env`
  - Set **CLIENT_ID** to the client ID of your application
  - Set the **APP_SECRET** you created

### Src folder

> Create a folder named **src** inside of the **server** folder. Inside of the new folder, create:

Reviewer 2: You can skip this since we are not following a specific guide. You can just something like

```
Here is a reference implementation of an express server:
```

Add the following code to the file:

```ts
import dotenv from 'dotenv';
import express from 'express';
import path from 'path';
import { getAccessTokenOnBehalfOf, validateJwt } from './auth';

var cors = require('cors');

// Load .env file
dotenv.config();

const app = express();
const PORT = process.env.port || process.env.PORT || 8000;

// Support JSON payloads
app.use(express.json());
app.use(express.static(path.join(__dirname, 'client')));

// enable cors for localhost for this sample
app.use(cors());

// An example for using POST and with token validation using middleware
app.post('/api/token', validateJwt, async (req, res) => {
  await getAccessTokenOnBehalfOf(req, res);
});

app.listen(PORT, () => {
  console.log(`⚡️[server]: Server is running at http://localhost:${PORT}`);
});
```

## Initialize the Teams Msal2 Provider

The Microsoft Graph Toolkit providers enable authentication and access to Microsoft Graph for the components. To learn more, see [Using the providers](../providers/providers.md). The [Teams Msal2 Provider](../providers/teams-msal2.md) handles all of the logic and interactions that need to be implemented with the Teams SDK to authenticate the user.

> For SSO-mode, make sure to provide `sso-url` / `ssoUrl` and have it point to you backend API.

Reviewer: ```suggestion
For SSO-mode, make sure to provide `sso-url` / `ssoUrl` and have it point to your backend API.
```

> ### Initialize in HTML

Reviewer: Another place where the tab format may be useful

Add the `mgt-teams-sso-provider` into the HTML `<body>`

```html
<mgt-teams-sso-provider 
  client-id="<YOUR_CLIENT_ID>"

>   auth-popup-url="/teamsauth"

Reviewer 2: Above you specify `auth.html`

```suggestion
  auth-popup-url="/auth.html"
```

  scopes="user.read,user.read.all,mail.readBasic,people.read.all,sites.read.all,user.readbasic.all,contacts.read,presence.read,presence.read.all,tasks.readwrite,tasks.read"
  sso-url="http://localhost:8000/api/token"
  http-method="POST"
  ></mgt-teams-sso-provider>
```

Replace `<YOUR_CLIENT_ID>` with the client ID for your application. 

### Initialize in JavaScript

To initialize the provider in your JavaScript code, import the Teams Provider and initialize the provider.

```ts
import {Providers} from '@microsoft/mgt-element';
import {TeamsSSOProvider, HttpMethod} from '@microsoft/mgt-teams-sso-provider';
import * as MicrosoftTeams from "@microsoft/teams-js";

TeamsSSOProvider.microsoftTeamsLib = MicrosoftTeams;

Providers.globalProvider = new TeamsSSOProvider({
  clientId: `<YOUR_CLIENT_ID>`,

>   authPopupUrl: '/teamsauth',

Reviewer 2: ```suggestion
  authPopupUrl: '/auth.html',
```

  scopes: ['user.read','user.read.all','mail.readBasic','people.read.all','sites.read.all','user.readbasic.all','contacts.read','presence.read','presence.read.all','tasks.readwrite','tasks.read'],
  ssoUrl: 'http://localhost:8000/api/token',
  httpMethod: HttpMethod.POST
});
```
Replace `<YOUR_CLIENT_ID>` with the client ID for your application.

> ## Add components

Reviewer: Format into tabs here as well
